tools/reviewer/web/src/app/dashboard/reviews-panel/review/review-titlebar/review-titlebar.component.ts

```diff
@@ -1,18 +1,53 @@
-import { Diff } from '@/shared';
-import { Component, EventEmitter, Input, Output } from '@angular/core';
+import { Component, Input, OnInit } from '@angular/core';
+
+import {
+  AuthService,
+  FirebaseService,
+  NotificationService
+} from '@/shared/services';
+import { Diff } from '@/shared/shell';
 
 @Component({
   selector: 'review-titlebar',
   templateUrl: './review-titlebar.component.html',
   styleUrls: ['./review-titlebar.component.scss']
 })
-export class ReviewTitlebarComponent {
-  @Input() diff: Diff.AsObject;
+export class ReviewTitlebarComponent implements OnInit {
+  statusList = [
+    'Review Not Started',
+    'Needs More Work',
+    'Under Review',
+    'Accepted',
+    'Submitting',
+    'Submitted',
+    'Reverting',
+    'Reverted',
+  ];
+  isLoading: boolean = true;
+  @Input() diff: Diff;
   @Input() editable;
-  @Output() onAddToAttentionList = new EventEmitter<string>();
 
-  // Save needAttentionOf list
-  saveAttentionList(name: string): void {
-    this.onAddToAttentionList.emit(name);
+  constructor(
+    public authService: AuthService,
+    private firebaseService: FirebaseService,
+    private notificationService: NotificationService,
+  ) { }
+
+  ngOnInit() {
+    this.isLoading = false;
+  }
+
+  updateAuthor(): void {
+    // Update need attention of the author in the diff
+    const author = this.diff.getAuthor();
+    author.setNeedsattention(!author.getNeedsattention());
+    this.diff.setAuthor(author);
+
+    // Send updated diff to firebase
+    this.firebaseService.updateDiff(this.diff).subscribe(() => {
```

Comment:
updateDiff is already present in firebaseService. All children can use it instead.
Gold rule: If you can put a reusable code to service, then you definitely should do it.
Use component binding for communication between parent and child only.

Reply:
One of the basic concepts off OOP is separation of concern. review-titlebar.component shouldn't be updating the diff even if it's calling a service method. It should only do stuff a titlebar is supposed to do, that is telling it's parent about a change. It's parent should be updating the diff. That's what I'm asking.

Reply:
@MateMahinKhan where do you see identical code at many places?
Also, my personal opinion about OOP concepts - they sometimes lead to more complicated code. There's some important properties of code (correctness, performance, simplicity, consistent style, maintainability, testability etc). These concepts are important also for non-OOP code, and IMO are more important than adhering to OOP concepts.

In this case, if the review-titlebar.component needs to update the diff, but it shouldn't (to adhere to a basic OOP concept), and instead it should call the parent, which will have the method that does this, then - 
* If we just call the method `updateDiff()` we aren't really doing much except some method indirection. It's pretty clear to anyone reading the code that Titlebar is updating the diff.
* If we don't call it `updateDiff()` (or similar), then it needs to be some abstraction of the action taking place. For example - `handleAuthorUpdate()` (but then you might need also `handleReviewerUpdate()`). So maybe more abstract - `handleTitleBarAction()` or something like that, so the Titlebar doesn't know this, but the diff is actually being saved. To me, although this is correct in one sense, Titlebar doesn't deal with saving Diff directly, it makes the code less simple, while not providing anything very useful for the other properties (correctness, performance etc).

So, in summary, IMO concepts like separation of concern, DRY and others (e.g http://java-design-patterns.com/principles), are all kind of ok, but can sometimes lead you to code that's actually worse in some aspects, and not better (or marginally better) in others.

Reply:
@oferb It was a method of updating the diff. 
Yeah, I'm of the strong opinion that titlebar shouldn't know about saving the diff. Angular 2 has made it really easy to make components and separate code into concerns and follow  I'm not sure but I guess it will result in better testable code as well. I see your point that it makes code complicated. 

Vadim, recently, started using review service that does the job as well of only one point of updating the diff from any child component.

Reply:
@MateMahinKhan I agree that Angular has some pretty good architecture guidelines for separating things correctly. All I'm saying is that we can choose whether to do that, and in this case, I don't think it will bring much value, since we aren't really testing the code.

As for tests - refactoring his code later to call the parent component would be fairly simple, so we can do that when we need to, for tests. Until then - we get simpler code.

Because we can refactor later, I view this as a reversible decision (AKA type 2 decision http://www.businessinsider.com/jeff-bezos-on-type-1-and-type-2-decisions-2016-4). There are decisions on testability that are type 1 decisions, like choosing a framework that is more testable (AFAIU Angular is more testable than React).